### ISSUE TITLE
Implicitly typed variables of types which were not annotateable in C# 8 aren't nullable (unlike all others)

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#50653]
**Version Used**: 
```
Branch master (19 Jan 2021)
Latest commit c36b5d3 by msftbot[bot]:
Merge pull request #50583 from CyrusNajmabadi/completoinCommon

Extract common code in completion providers.
```

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
class C
{
  void M1<T>(T t)
  {
    var x = t; // type of 'x' is 'T' instead of 'T?'
    x = default; // redundant CS8600 in C# 9, no warnings in C# 8
  }
    
  void M2<T>(T t) where T: class
  {
    var x = t; // type of 'x' is 'T?'
    x = default; // no warnings
  }
}
```

**Expected Behavior**:
No warnings for `x = default` assignment in both methods

**Actual Behavior**:
`CS8600: Converting null literal or possible null value to non-nullable type.` is reported in `M1`

**Notes**
There's an LDM decision that `var` variables should be nullable by default because otherwise people will have to specify exact type just because they might assign `null` values later which is preventable by flow analysis: https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-12-18.md#var

In C# 8 `T?` type was unspeakable and couldn't be used directly. Due to this reason Roslyn used unannotated version of the type for `var` variables. However Roslyn allowed `default` or `null` assignments to such types without warnings because it was impossible to declare a null-allowing version of such types. The code in example above _did not have any warnings in C# 8_.

In C# 9 however the `T?` syntax is allowed and Roslyn started to warn about `default` or `null` values assigned to `T` variables. This also introduced additional warnings for `var` variables since their annotations weren't updated with this change even though the `T?` type became perfectly valid in this lang version.

Since the `T?` syntax became valid the same LDM reasons that led to `var` being nullable are now applicable to type that didn't allow annotations in C# 8. It looks like there's no reasons left to treat these types differently. It's also important to note that it this issue also adds some unnecessary warnings when updating lang version to C# 9

/cc @jaredpar

**Comment by RikkiGibson** [comment]
I'm wondering if we should make the type of the `var` dependent on LangVersion, or if we should always make it of type `T?`. It feels like if it were always ` T?` then such types could slip into implicit type arguments..but I think that may already happen based on the flow state of the local variable.

Also tagging @cston

**Comment by TessenR** [comment]
@RikkiGibson Yes, this already happens giving a quite confusing warning in C# 8

Compile the following code in C# 8:
```
#nullable enable
class C<T>
{
  void M2()
  {
    T t = default;
    // warning CS8619: Nullability of reference types in value of type 'C<T?>' doesn't match target type 'C<T>'.
    C<T> c = GetC(t);
  }

  C<T> GetC<T>(T t) => new C<T>();
}
```
Note that Roslyn complains about `C<T?>` type which is not speakable in C# 8

**Comment by jaredpar** [comment]
> I'm wondering if we should make the type of the var dependent on LangVersion, or if we should always make it of type T?. 

The decision in LDM was to always allow assignment to a `var` when the flow state was "maybe-null". The decision was to make it effective for C# 8 immediately because it was an adoption blocker for 8.  Later we did make some changes conditional on the language version but I'm not sure it's worth going back and doing this.

**Comment by RikkiGibson** [comment]
I think we should probably address this by making 'var' use an always annotated type for reference types, in all language versions. This makes "maybe-default" always assignable to the var without warnings.

I think the effect in practice will only be to remove warnings. Can we try to do this in 16.9? Perhaps it's not urgent enough but it feels like it would be nice to get the change to users sooner if it's low impact in practice.

**Comment by jaredpar** [comment]
Pretty sure we discussed this case in LDM, it's a case I'm interested in hence fairly confident I brought it up. Believe the discussion pushed towards this being by design though. 

@cston as he implemented this and can speak to the design aspect here.
